src/Swarm/TUI/Controller.hs

```diff
@@ -587,76 +587,96 @@ stripCmd pty = pty
+            . (gameState %~ execState (activateRobot 0))
+
+    if not $ s ^. gameState . replWorking
+      then continue $ case entry of
+        CmdPrompt uinput _ ->
+          case processTerm' topTypeCtx topCapCtx uinput of
+            Right mt ->
+              maybe id startBaseProgram mt
+                . (uiState . uiReplHistory %~ addREPLItem (REPLEntry uinput))
+                . (uiState %~ resetWithREPLForm (set promptUpdateL "" (s ^. uiState)))
+                $ s
+            Left err -> s & uiState . uiError ?~ err
+        SearchPrompt t hist ->
+          case lastEntry t hist of
+            Nothing -> s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
+            Just found
+              | T.null t -> s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
+              | otherwise ->
+                s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt found)
+                  & validateREPLForm
+      else continueWithoutRedraw s
+  Key V.KUp -> continue $ s & adjReplHistIndex Older
+  Key V.KDown -> continue $ s & adjReplHistIndex Newer
+  ControlKey 'r' -> continue $
+    case s ^. uiState . uiReplForm . to formState of
+      CmdPrompt uinput _ ->
+        let newform = mkReplForm $ SearchPrompt uinput (s ^. uiState . uiReplHistory)
          in s & uiState . uiReplForm .~ newform
+      SearchPrompt ftext rh -> case lastEntry ftext rh of
+        Nothing -> s
+        Just found ->
+          let newform = mkReplForm $ SearchPrompt ftext (removeEntry found rh)
+           in s & uiState . uiReplForm .~ newform
+  CharKey '\t' -> do
+    let formSt = s ^. uiState . uiReplForm . to formState
+        newform = mkReplForm $ tabComplete s formSt
+    continue $
+      s & uiState . uiReplForm .~ newform
+        & validateREPLForm
+  EscapeKey ->
+    case s ^. uiState . uiReplForm . to formState of
+      CmdPrompt {} -> continueWithoutRedraw s
+      SearchPrompt _ _ ->
+        continue $ s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
+  ev -> do
+    f' <- handleFormEvent ev (s ^. uiState . uiReplForm)
+    continue $
+      case formState f' of
+        CmdPrompt {} -> validateREPLForm (s & uiState . uiReplForm .~ f')
+        SearchPrompt t _ ->
+          let newform = set promptUpdateL t (s ^. uiState)
+           in s & uiState . uiReplForm .~ newform
+
+-- | Try to complete the last word in a partially entered REPL prompt using
+--   things reserved words and names in scope.
+tabComplete :: AppState -> REPLPrompt -> REPLPrompt
+tabComplete _ p@(SearchPrompt {}) = p
+tabComplete _ (CmdPrompt "" []) = CmdPrompt "" []
+tabComplete s (CmdPrompt t []) = case matches of
+  [] -> CmdPrompt t []
+  [m] -> CmdPrompt (completeWith m) []
+  (m : ms) -> CmdPrompt (completeWith m) (ms ++ [m])
+ where
+  completeWith m = T.append t (T.drop (T.length lastWord) m)
+  lastWord = T.takeWhileEnd isIdentChar t
```

Comment:
This does not work for completing operators, e.g. `=` get completed as `=noop`.

Maybe there should be a check that before the "last word", there is either nothing or a whitespace character?

Reply:
Thanks, filed as #518 .